toplevel/topdirs.ml

```diff
@@ -401,6 +401,52 @@ let _ = add_directive "remove_printer"
       doc = "Remove the named function from the table of toplevel printers.";
     }
 
+(* PPX-related stuff *)
+
+(* Installing mapper specified by indent if it exists and type is appropriate *)
+let install_ppx_mapper ppf (lid: Longident.t) =
+  try
+    let (path, desc) = Env.lookup_value lid !toplevel_env in
+    let open Path in
+    let rec is_good_type t =
```

Comment:
This seems a rather fragile way to match on types.  You should do as for printer registration (see e.g. `match_simple_printer_type`), using proper unification.